gollm/gemini.go

```diff
@@ -114,111 +107,98 @@ func (c *GeminiClient) SetResponseSchema(responseSchema *Schema) error {
+	var config *genai.GenerateContentConfig
 
 	if c.responseSchema != nil {
-		model.ResponseSchema = c.responseSchema
-		model.ResponseMIMEType = "application/json"
+		config = &genai.GenerateContentConfig{
+			ResponseSchema:   c.responseSchema,
+			ResponseMIMEType: "application/json",
+		}
 	}
 
-	var geminiParts []genai.Part
-
-	geminiParts = append(geminiParts, genai.Text(request.Prompt))
+	content := []*genai.Content{
+		{Role: "user", Parts: []*genai.Part{{Text: request.Prompt}}},
+	}
 
-	log.Info("sending GenerateContent request to gemini", "parts", geminiParts)
-	geminiResponse, err := model.GenerateContent(ctx, geminiParts...)
+	log.Info("sending GenerateContent request to gemini", "content", content)
+	result, err := c.client.Models.GenerateContent(ctx, c.model, content, config)
 	if err != nil {
 		return nil, err
 	}
 
-	if len(geminiResponse.Candidates) == 0 {
-		return nil, fmt.Errorf("got no responses from gemini")
-	}
-
-	if len(geminiResponse.Candidates) > 1 {
-		log.Info("only considering first candidate")
-		for i := 1; i < len(geminiResponse.Candidates); i++ {
-			candidate := geminiResponse.Candidates[i]
-			log.Info("ignoring candidate: %q", candidate.Content)
-		}
-	}
-	var response strings.Builder
-	candidate := geminiResponse.Candidates[0]
-	for _, part := range candidate.Content.Parts {
-		switch part := part.(type) {
-		case genai.Text:
-			if response.Len() != 0 {
-				response.WriteString("\n")
-			}
-			response.WriteString(string(part))
-		default:
-			return nil, fmt.Errorf("unexpected type of content part: %T", part)
-		}
-	}
-
-	return &GeminiCompletionResponse{geminiResponse: geminiResponse, text: response.String()}, nil
+	return &GeminiCompletionResponse{geminiResponse: result, text: result.Text()}, nil
 }
 
 // StartChat starts a new chat with the model.
 func (c *GeminiClient) StartChat(systemPrompt string) Chat {
-	model := c.client.GenerativeModel(c.model)
-
 	// Some values that are recommended by aistudio
-	model.SetTemperature(1)
-	model.SetTopK(40)
-	model.SetTopP(0.95)
-	model.SetMaxOutputTokens(8192)
-	model.ResponseMIMEType = "text/plain"
-
-	if c.responseSchema != nil {
-		model.ResponseSchema = c.responseSchema
-		model.ResponseMIMEType = "application/json"
+	temperature := float32(1.0)
+	topK := float32(40)
+	topP := float32(0.95)
+	maxOutputTokens := int32(8192)
+
+	chat := &GeminiChat{
+		model:  c.model,
+		client: c.client,
+		genConfig: &genai.GenerateContentConfig{
+			SystemInstruction: &genai.Content{
+				Parts: []*genai.Part{
+					{Text: systemPrompt},
+				},
+			},
+			Temperature:      &temperature,
+			TopK:             &topK,
+			TopP:             &topP,
+			MaxOutputTokens:  &maxOutputTokens,
+			ResponseMIMEType: "text/plain",
+		},
+		history: []*genai.Content{},
 	}
 
-	if systemPrompt != "" {
-		model.SystemInstruction = &genai.Content{
-			Parts: []genai.Part{
-				genai.Text(systemPrompt),
-			},
+	if c.model == "gemma-3-27b-it" {
```

Comment:
Interesting .... TBD whether this belongs here or in the caller.  But let's start with this and see how we feel over time.

Reply:
Ack.